include/CObjectCache.h

```diff
@@ -70,191 +122,433 @@ namespace impl
+    >
+    struct CMultiObjectCacheBase;
 
-    inline bool insert(const K& _key, T* _val)
+    template<
+        template<typename...> class ContainerT_T,
+        typename T, //value type for container
+        typename ...K //optionally key type for std::map/std::unordered_map
+    >
+    struct CMultiObjectCacheBase<true, ContainerT_T, T, K...> : public CObjectCacheBase<ContainerT_T, T, K...>, public CMultiCache_tag
     {
-        auto it = find(_key);
-        if (it != std::end(m_container))
-            return false;
+    private:
+        using Base = CObjectCacheBase<ContainerT_T, T, K...>;
 
-        greet(_val);
-        m_container.insert(it, {_key, _val});
-        return true;
-    }
+    public:
+        using Base::Base;
 
-    //! Returns true if had to insert
-    inline bool swapValue(const K& _key, T* _val)
+        inline bool insert(const typename Base::KeyType_impl& _key, typename Base::ValueType_impl _val)
+        {
+            INSERT_IMPL_ASSOC
+        }
+
+        inline typename Base::RangeType findRange(const typename Base::KeyType_impl& _key)
+        {
+            return Base::m_container.equal_range(_key);
+        }
+        inline typename Base::ConstRangeType findRange(const typename Base::KeyType_impl& _key) const
+        {
+            return Base::m_container.equal_range(_key);
+        }
+    };
+    template<
+        template<typename...> class ContainerT_T,
+        typename T, //value type for container
+        typename ...K //optionally key type for std::map/std::unordered_map
+    >
+    struct CMultiObjectCacheBase<false, ContainerT_T, T, K...> : public CObjectCacheBase<ContainerT_T, T, K...>, public CMultiCache_tag
     {
-        greet(_val); //grab before drop
+    private:
+        using Base = CObjectCacheBase<ContainerT_T, T, K...>;
+
+    public:
+        using Base::Base;
 
-        auto it = find(_key);
-        if (it != std::end(m_container))
+        inline bool insert(const typename Base::KeyType_impl& _key, typename Base::ValueType_impl _val)
         {
-            dispose(it->second);
-            it->second = _val;
-            return false;
+            INSERT_IMPL_VEC
         }
 
-        m_container.insert(it, {_key, _val});
-        return true;
-    }
+    private:
+        template<typename RngType>
+        inline RngType findRange_internal(const typename Base::KeyType_impl& _key)
+        {
+            auto cmpf = [](const typename Base::PairType_impl& _a, const typename Base::PairType_impl& _b) -> bool {return _a.first < _b.first; };
+            typename Base::PairType_impl lookingFor{_key, nullptr};
+
+            RngType range;
+            range.first = std::lower_bound(std::begin(Base::m_container), std::end(Base::m_container), lookingFor, cmpf);
+            if (range.first == std::end(Base::m_container) || _key < range.first->first)
+            {
+                range.second = range.first;
```

Comment:
if you retuned std::end() for first and second then any function under a lock (or single threaded) would not have to use `isNonZeroRange` to determine if valid range was found

Reply:
But it wouldn't give any information where to put new element/relation between contents of cache and elent being inserted. `findRange` is used also in functions where insertions happens

Reply:
ok I guess thats fine with using != for isNonZeroRange